### ISSUE TITLE
TypeScript 3.7.x: Optional chaining and nullish coalescing is broken

### ISSUE THREAD

**Issue by smithki** [developit/microbundle#522]
The latest version of TypeScript (`3.7.x`) is broken if using optional chaining and nullish coalescing. It fails with the following error:

```sh
SyntaxError: /path/to/file.ts: Support for the experimental syntax 'optionalChaining' isn't currently enabled

...

Add @babel/plugin-proposal-optional-chaining (https://git.io/vb4Sk) to the 'plugins' section of your Babel config to enable transformation.
```

The simple fix (I think) is to update the TypeScript dependency in `microbundle`. _However..._ my preference would be for `microbundle` to detect if I have another version of TypeScript already registered in my `package.json` and _use that version_ as the default (falling back to the included dependency if required).

**Comment by smithki** [comment]
I spoke too soon. After digging a little deeper, it seems the issue is actually related to TypeScript emitting optional chaining and nullish coalescing without transpilation with `esnext`... Until the babel plugin above is added to `microbundle`, is there a local workaround I can use? It doesn't appear that I can extend the babel config...

I updated the title to reflect this new information.

**Comment by tunnckoCore** [comment]
> to detect if I have another version of TypeScript already registered in my

Microbundle doesn't do anything special, it uses the typescript plugin which is responsible for handling that.

> is there a local workaround I can use?

I guess switch the `esnext`? I'm not sure.

**Comment by ForsakenHarmony** [comment]
You can add a custom babelrc (the first babel plugin is for constants)

**Comment by smithki** [comment]
I figured out from reading the source that a a custom babelrc is supported in the next version. I think this can be closed now... though I wonder, is it out of scope for `microbundle` to allow TypeScript to handle all transpiration? My overall preference would be for `microbundle` to respect the value of `target` in my `tsconfig.json`... is there a reason why that is not the case?

**Comment by ConsoleTVs** [comment]
I came to the same conclusion. Shame, it was my first shot on this lib. It's a pretty it does not merge babel settings or at least, allow you to customize it or append plugins. Anyway, this should be a top priority because as for today, the statement: **Typescript supported**, it's not true. Will keep an eye here for the fix. And if the fix is already there in the new version, why not publish it? @developit

**Comment by ForsakenHarmony** [comment]
@smithki I'd assume you already are on the next version, because we didn't have babel in 0.11.0

It should work if you add a `.babelrc` to your project with this content

```json
{
	"plugins": [
		[
			"@babel/plugin-proposal-optional-chaining"
		]
	]
}

```

**Comment by smithki** [comment]
@ForsakenHarmony yeah, I've managed to get what I need in `microbundle@next`, but unfortunately for the use-case I'm after, the rest of my team is wary of pre-release NPM packages. When will `0.12.x` be released mainstream? It would be a lifesaver.

**Comment by Albert-Gao** [comment]
adding `.babelrc` does not work for me, using `microbundle@next` solves the issue

**Comment by ForsakenHarmony** [comment]
I mentioned that in my commment

**Comment by adjohnston** [comment]
> @smithki I'd assume you already are on the next version, because we didn't have babel in 0.11.0
> 
> It should work if you add a `.babelrc` to your project with this content (and install the package)
> 
> ```json
> {
> 	"plugins": [
> 		[
> 			"@babel/plugin-proposal-optional-chaining"
> 		]
> 	]
> }
> ```

Do the above worked for me in the newly released v0.12.0.

**Comment by unlight** [comment]
microbundle, 0.12.0-next.9

```
export async function main(x) {
    console.log(x ?? (await '1'));
}
```

Produces rubbish
```ts
const main = function (x) {
  try {
    function _temp(_) {
      console.log(_);
    }

    return Promise.resolve(x ? Promise.resolve('1').then(_temp) : _temp(x));
  } catch (e) {
    return Promise.reject(e);
  }
};
```

Adding `.babelrc` did not help.

**Comment by TechQuery** [comment]
This bug haven't been fixed in version `0.12.2`, and a new error is thrown out:
> (terser plugin) SyntaxError: Unexpected token: punc (.)

**Comment by ConsoleTVs** [comment]
Just leaving this here in case some of you have issues with the compression of typescript.

- Add the nessesary scripts to the `package.json` file

```json
"scripts": {
  "dev": "microbundle",
  "prod": "microbundle --compress",
  "watch": "microbundle watch"
},
```

- Install the following packages babel packages and add them to the .babelrc file in the root of your project:

```json
{
  "plugins": [
    "@babel/plugin-transform-typescript",
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "@babel/plugin-proposal-object-rest-spread"
  ]
}
```

**Comment by TechQuery** [comment]
@ConsoleTVs I got a new error
> (babel plugin) Error: .plugins[0][1] must be an object, false, or undefined

with configuration of
```json
{
    "plugins": [
        [
            "@babel/plugin-transform-typescript",
            "@babel/plugin-proposal-optional-chaining"
        ]
    ]
}
```

**Comment by ConsoleTVs** [comment]
My example is located here: https://github.com/Chartisan/Chartjs

inside the `.babelrc` file.

**Comment by lukas-valenta** [comment]
@TechQuery you have an error in the config, it should be:
```
{
  "plugins": [
    "@babel/plugin-transform-typescript",
    "@babel/plugin-proposal-optional-chaining"    
  ]
}
```

or

```
{
  "plugins": [
    ["@babel/plugin-transform-typescript"],
    ["@babel/plugin-proposal-optional-chaining]"    
  ]
}
```

**Comment by TechQuery** [comment]
> This bug haven't been fixed in version `0.12.2`, and a new error is thrown out:
> 
> > (terser plugin) SyntaxError: Unexpected token: punc (.)

@lukas-valenta I sorry for one more `[ ]`, but the following configuration gets the same error like before:
```json
{
    "plugins": [
        "@babel/plugin-transform-typescript",
        "@babel/plugin-proposal-optional-chaining"
    ]
}
```

**Comment by ConsoleTVs** [comment]
dude I commented here for a reason. You need those 4:

```
"@babel/plugin-transform-typescript",
"@babel/plugin-proposal-optional-chaining",
"@babel/plugin-proposal-nullish-coalescing-operator",
"@babel/plugin-proposal-object-rest-spread"
```

**Comment by TechQuery** [comment]
@ConsoleTVs I find the reason isn't the 2 more plugins, just because `microbundle` don't process TypeScript files in `node_modules`, so we need set them into `--external`.

**Comment by yuschick** [comment]
> > @smithki I'd assume you already are on the next version, because we didn't have babel in 0.11.0
> > It should work if you add a `.babelrc` to your project with this content (and install the package)
> > ```json
> > {
> > 	"plugins": [
> > 		[
> > 			"@babel/plugin-proposal-optional-chaining"
> > 		]
> > 	]
> > }
> > ```
> 
> Doing the above worked for me in the newly released v0.12.0.

I am running Microbundle v0.12.3 and this solution worked for me.
